app/src/main/java/net/squanchy/schedule/tracksfilter/ScheduleTracksFilter.kt

```diff
@@ -77,41 +89,51 @@ class ScheduleTracksFilterActivity : AppCompatActivity() {
     }
 }
 
-private class TracksFilterAdapter(context: Context) : RecyclerView.Adapter<TrackViewHolder>() {
+private class TracksFilterAdapter(
+    context: Context,
+    private val trackStateChangeListener: OnTrackSelectedChangeListener
+) : ListAdapter<CheckableTrack, TrackViewHolder>(DiffCallback()) {
 
     init {
         setHasStableIds(true)
     }
 
     private val layoutInflater = LayoutInflater.from(context)
 
-    private var checkableTracks: List<CheckableTrack> = emptyList()
-    private lateinit var trackStateChangeListener: OnTrackSelectedChangeListener
-
-    override fun getItemCount(): Int = checkableTracks.size
-
     override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TrackViewHolder {
         val view = layoutInflater.inflate(R.layout.track_filters_item, parent, false) as CheckBox
         return TrackViewHolder(view)
     }
 
     override fun onBindViewHolder(holder: TrackViewHolder, position: Int) {
-        holder.bind(checkableTracks[position], trackStateChangeListener)
+        holder.bind(getItem(position), trackStateChangeListener)
     }
 
     override fun getItemId(position: Int): Long {
-        return checkableTracks[position].track().id.hashCode().toLong() // TODO this should use a proper checksum
+        return getItem(position).track().id.hashCode().toLong() // TODO this should use a proper checksum
     }
 
-    fun updateTracks(newCheckableTracks: List<CheckableTrack>, listener: OnTrackSelectedChangeListener) {
-        checkableTracks = newCheckableTracks // TODO use DiffUtil instead
-        trackStateChangeListener = listener
-        notifyDataSetChanged()
+    class DiffCallback : DiffUtil.ItemCallback<CheckableTrack>() {
+        override fun areItemsTheSame(oldItem: CheckableTrack?, newItem: CheckableTrack?): Boolean {
```

Comment:
you might be able to remove the `?` (probably added because kotlin doesn't know if they're nullable)

Reply:
No, in this callback they are actually nullable, I had this biting me in the past (it's the `DiffUtils` callback)